### PR TITLE
mixmrt/i2s: Add I2S protocol support.

### PR DESCRIPTION
**Pull request by miketeachman**
- I2S API consistent with STM32, ESP32, and RP2 ports
- I2S features:
  - controller transmit and controller receive
  - 16-bit and 32-bit sample sizes
  - mono and stereo formats
  - sampling frequencies from 8kHz to 48kHz
  - 3 modes of operation:
    - blocking
    - non-blocking with callback
    - uasyncio
  - configurable internal buffer
  - optional MCK
- tested with the following development boards:
  - MIMXRT1010_EVK, MIMXRT1015_EVK, MIMXRT1020_EVK, MIMXRT1050_EVK
  - Teensy 4.0, Teensy 4.1
  - Olimex RT1010
  - Seeed ARCH MIX
- tested with the following I2S hardware peripherals:
  - UDA1334
  - GY-SPH0645LM4H
  - WM8960 codec on board the MIMXRT boards and separate breakout board
  - INMP441
  - PCM5102
  - SGTL5000 on the Teensy audio shield
- mimxrt build metric changes versus master branch: text(+7464), data(+28), bss(+128)
- mimxrt quickref doc updates are missing and are dependent on https://github.com/micropython/micropython/pull/7494 being accepted.
- this PR is a collaborative effort between myself, @robert-hh, @alphaFred, and @rdagger

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for this, it looks good!  I tested it on a Teensy 4.0 with the new `machine_i2s_rate.py` test and it works.

Merged in 1f6cb8f0476e699c1910ed59f1bb027fbef72ad7

File: ports/mimxrt/pin.c [hunk 1]
```diff
@@ -171,7 +171,7 @@ const machine_pin_obj_t *pin_find(mp_obj_t user_obj) {
         return pin_obj;
     }
 
-    mp_raise_ValueError(MP_ERROR_TEXT("Pin doesn't exist"));
+    mp_raise_ValueError(MP_ERROR_TEXT("Pin does not exist"));
```
Comment (dpgeorge) [inline ports/mimxrt/pin.c:174]: I reverted this change, it should use contracted form to match everywhere else.

Comment (miketeachman) [reply ports/mimxrt/pin.c:174]: The intent was to be more easily understood by users who have English as a 2nd language.   I see the point about consistency.

Comment (dpgeorge) [reply ports/mimxrt/pin.c:174]: There was a commit/PR from a while ago that went through and contracted everything, with the main aim to reduce code size.

**Comment by miketeachman** [comment]
> Thanks for this, it looks good! I tested it on a Teensy 4.0 with the new `machine_i2s_rate.py` test and it works.
> 
> Merged in [1f6cb8f](https://github.com/micropython/micropython/commit/1f6cb8f0476e699c1910ed59f1bb027fbef72ad7)

Great!  Thank you for pulling this in.  I'll keep an eye out for the [doc PR ](https://github.com/micropython/micropython/pull/7494) and then update the quickref for I2S

**Pull request closed without merging**